### PR TITLE
Annotate System.Transactions.Local for nullable

### PR DESCRIPTION
**Pull request by buyaa-n**
Contributes to https://github.com/dotnet/corefx/issues/40623

### CODE REVIEW

File: src/System.Transactions.Local/src/System/Transactions/TransactionState.cs [hunk 1]
```diff
@@ -3672,7 +3688,7 @@ internal override byte[] PromotedToken(InternalTransaction tx)
 
         internal override void DisposeRoot(InternalTransaction tx)
         {
-            tx.State.Rollback(tx, null);
+            tx.State!.Rollback(tx, null);
```
Comment (buyaa-n) [inline src/System.Transactions.Local/src/System/Transactions/TransactionState.cs]: Banged references like this are all set or asserted in the calling context

File: src/System.Transactions.Local/src/System/Transactions/DependentTransaction.cs [hunk 2]
```diff
@@ -3,6 +3,8 @@
 // See the LICENSE file in the project root for more information.
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/DependentTransaction.cs]: Nit, not introduced by your change, but there is two empty lines here.

File: src/System.Transactions.Local/src/System/Transactions/EnlistmentTraceIdentifier.cs [hunk 3]
```diff
@@ -47,7 +47,7 @@ namespace System.Transactions
 
         public override int GetHashCode() => base.GetHashCode();  // Don't have anything better to do.
 
-        public override bool Equals(object obj) => obj is EnlistmentTraceIdentifier && Equals((EnlistmentTraceIdentifier)obj);
+        public override bool Equals(object? obj) => obj is EnlistmentTraceIdentifier && Equals((EnlistmentTraceIdentifier)obj);
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/EnlistmentTraceIdentifier.cs]: We could avoid an extra cast here:

```cs
public override bool Equals(object? obj) => obj is EnlistmentTraceIdentifier enlistmentTraceId && Equals(enlistmentTraceId);
```

File: src/System.Transactions.Local/src/System/Transactions/Transaction.cs [hunk 4]
```diff
@@ -319,9 +317,9 @@ public override int GetHashCode()
 
         // Don't allow equals to get the identifier
         //
-        public override bool Equals(object obj)
+        public override bool Equals(object? obj)
         {
-            Transaction transaction = obj as Transaction;
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/Transaction.cs]: Nit: this method method could be simplified to a single return expression if using the `is` operator.

File: src/System.Transactions.Local/src/System/Transactions/TransactionException.cs [hunk 5]
```diff
@@ -112,25 +112,25 @@ internal static TransactionException Create(string message, Guid distributedTxId
...
-            string messagewithTxId = message;
+            string? messagewithTxId = message;
             if (IncludeDistributedTxId(distributedTxId))
                 messagewithTxId = SR.Format(SR.DistributedTxIDInTransactionException, messagewithTxId, distributedTxId);
 
             return Create(messagewithTxId, innerException);
         }
 
-        internal static TransactionException Create(TraceSourceType traceSource, string message, Exception innerException, Guid distributedTxId)
+        internal static TransactionException Create(TraceSourceType traceSource, string? message, Exception innerException, Guid distributedTxId)
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/TransactionException.cs]: `innerException` should be nullable right?

File: src/System.Transactions.Local/src/System/Transactions/TransactionManager.cs [hunk 6]
```diff
@@ -63,8 +64,8 @@ internal static void ProcessExistingTransactions(TransactionStartedEventHandler
                 IDictionaryEnumerator e = PromotedTransactionTable.GetEnumerator();
                 while (e.MoveNext())
                 {
-                    WeakReference weakRef = (WeakReference)e.Value;
-                    Transaction tx = (Transaction)weakRef.Target;
+                    WeakReference weakRef = (WeakReference)e.Value!;
+                    Transaction? tx = (Transaction?)weakRef.Target;
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/TransactionManager.cs]: Nit: simplify to using `is`?

File: src/System.Transactions.Local/src/System/Transactions/TransactionManager.cs [hunk 7]
```diff
@@ -92,13 +93,13 @@ internal static void FireDistributedTransactionStarted(Transaction transaction)
...
         // Data storage for current delegate
-        internal static HostCurrentTransactionCallback s_currentDelegate = null;
+        internal static HostCurrentTransactionCallback? s_currentDelegate = null;
         internal static bool s_currentDelegateSet = false;
 
         // CurrentDelegate
         //
         // Store a delegate to be used to query for an external current transaction.
-        public static HostCurrentTransactionCallback HostCurrentCallback
+        [DisallowNull] public static HostCurrentTransactionCallback? HostCurrentCallback
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/TransactionManager.cs]: Please put attribute in a new line.

File: src/System.Transactions.Local/src/System/Transactions/TransactionOptions.cs [hunk 8]
```diff
@@ -23,7 +23,7 @@ public IsolationLevel IsolationLevel
 
         public override int GetHashCode() => base.GetHashCode();  // Don't have anything better to do.
 
-        public override bool Equals(object obj) => obj is TransactionOptions && Equals((TransactionOptions)obj);
+        public override bool Equals(object? obj) => obj is TransactionOptions && Equals((TransactionOptions)obj);
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/TransactionOptions.cs]: Can remove the cast if you use the already in use `is` operator to declare inline:

```suggestion
        public override bool Equals(object? obj) => obj is TransactionOptions transactionOptions && Equals(obj);
```

File: src/System.Transactions.Local/src/System/Transactions/TransactionTraceIdentifier.cs [hunk 9]
```diff
@@ -34,7 +34,7 @@ public TransactionTraceIdentifier(string transactionIdentifier, int cloneIdentif
 
         public override int GetHashCode() => base.GetHashCode();  // Don't have anything better to do.
 
-        public override bool Equals(object obj) => obj is TransactionTraceIdentifier && Equals((TransactionTraceIdentifier)obj);
+        public override bool Equals(object? obj) => obj is TransactionTraceIdentifier && Equals((TransactionTraceIdentifier)obj);
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/TransactionTraceIdentifier.cs]: Ditto: 

```suggestion
        public override bool Equals(object? obj) => obj is TransactionTraceIdentifier transactionTraceId && Equals(transactionTraceId);
```

**Review by safern — APPROVED** [review]

File: src/System.Transactions.Local/src/System/Transactions/TransactionException.cs (hunk 5 above)
Comment (buyaa-n) [reply src/System.Transactions.Local/src/System/Transactions/TransactionException.cs]: yes

File: src/System.Transactions.Local/src/System/Transactions/Transaction.cs [hunk 10]
```diff
@@ -319,17 +319,15 @@ public override int GetHashCode()
         //
         public override bool Equals(object? obj)
         {
-            Transaction? transaction = obj as Transaction;
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/Transaction.cs]: You could just do: 

```cs
return obj is Transaction transaction &&
    _internalTransaction.TransactionHash == transaction._internalTransaction.TransactionHash;
```

Comment (buyaa-n) [reply src/System.Transactions.Local/src/System/Transactions/Transaction.cs]: sure

File: src/System.Transactions.Local/src/System/Transactions/Transaction.cs [hunk 11]
```diff
@@ -319,37 +317,29 @@ public override int GetHashCode()
...
-            // to this, which is a Transaction.
-            if (null == transaction)
-            {
-                return false;
-            }
-
-            // Check the internal transaction object for equality.
-            return _internalTransaction.TransactionHash == transaction._internalTransaction.TransactionHash;
+            // to this, which is a Transaction. Check the internal transaction object for equality.
+            return obj is Transaction transaction &&_internalTransaction.TransactionHash == transaction._internalTransaction.TransactionHash;
```
Comment (safern) [inline src/System.Transactions.Local/src/System/Transactions/Transaction.cs]: Unfortunately, you missed a space after `&&`
```suggestion
            return obj is Transaction transaction && _internalTransaction.TransactionHash == transaction._internalTransaction.TransactionHash;
```

**Pull request merged**